### DISCUSSION TITLE
"Dockable" panels?

### DISCUSSION THREAD

**Discussion by mikesbytes** [slint-ui/slint#1386] [Ideas]
Hello,

Apologies if this has already been discussed but I couldn't find anything when searching.

A common UI pattern in desktop apps is to have various panels that can be toggled on and off, reorganized within the UI via drag and drop, and resized. Example applications would be pretty much anything made by Adobe, Krita, Inkscape, Blender, most IDEs etc.

Are there any plans to implement such systems? If not, is there some way to implement this pattern with the tools already existing in Slint? I have been skimming the documentation and I see some possible ways to accomplish this, however none of them are particularly clean.

**Comment by ogoffart** [comment]
This is not yet implemented, but surely something that we want to provide.
This is currently not in our short term roadmap, but we could help with implementation if somebody is interested.

We don't have yet the ability to create detachable window in slint, but inner window are possible.  
Example with this: https://slint-ui.com/releases/0.2.4/editor/?load_url=https://raw.githubusercontent.com/slint-ui/slint/master/examples/fancy_demo/main.slint
Now, there will be the need for some native code to adjust the sub-window (or docks) geometry based on some docking logic, and that should be possible.


With support from the slint compiler, we could imagine some API like this:

```
DockableArea {
  Dock { 
      title: "Foo";
      Rectangle { color: blue; /* ... / } 
  }
  Dock { 
      title: "Bar";
      Rectangle { color: red; /* ... */ } 
  }

  Rectangle { color: white; /* this is the main content */ }
}
```

**Reply by mikesbytes** [reply]
Thanks for the quick response! Good to hear this is something that is at least on the table. I would love to have something along the lines of [this Qt library](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System) eventually.

For now I think I can get away with just having a fixed config for the "panels" that can be edited with either a config file or GUI configurator. If I am able to come up with anything that enables drag and drop reorganization though, I will definitely update this thread with my results.

**Comment by bertlebee** [comment]
Any news on drag and drop support? I noticed it's used in the lsp preview but it looks like the code for this lives in the LSP code which I guess means it's not actually a slint feature at this stage, but it looks like the hard work may have been done?

**Reply by ogoffart** [reply]
The LSP don't really have dockable panel.
And if it had, you'd be able to implement them the same way in your application.

**Reply by bertlebee** [reply]
I was talking more generally about drag and drop support rather than dockable panels specifically, but I guess your point remains about being able to implement it in the same way.

**Comment by akvmamusic** [comment]
any news on this? should we just create our own implementations?

**Reply by ogoffart** [reply]
This feature is currently not on our short-term roadmap. We encourage you to take a shot at implementing it. 
(Our roadmap is flexible and responsive to the needs of our users and customers. If this issue is a blocker for you, please let us know by [email](mailto:info@slint.dev) or [chat](https://chat.slint.dev/). Share details about your project, and we'll consider prioritizing it or offering it as a service.)
